xla/python/py_array.cc

```diff
@@ -629,11 +637,68 @@ StatusOr<std::uintptr_t> PyArray::UnsafeBufferPointer() {
       IfrtHelpers::pjrt_buffer(arr.ifrt_array()));
 }
 
-StatusOr<py::dict> PyArray::CudaArrayInterface() {
-  TF_ASSIGN_OR_RETURN(auto arr, AssertUnsharded("UnsafeBufferPointer"));
+py::dict PyArray::CudaArrayInterface() {
+  auto arr = ValueOrThrow(AssertUnsharded("UnsafeBufferPointer"));
+
+  ifrt::Array* ifrt_array = arr.ifrt_array();
+  std::optional<Shape>& scratch = arr.GetStorage().dynamic_shape;
+  auto* pjrt_buffer = IfrtHelpers::pjrt_buffer(ifrt_array);
+  if (pjrt_buffer->client()->platform_id() != CudaId()) {
+    throw py::attribute_error(
+        "__cuda_array_interface__ is only defined for NVidia GPU buffers.");
+  }
+  if (pjrt_buffer->IsTuple()) {
+    throw py::attribute_error(
+        "__cuda_array_interface__ is only defined for array buffers.");
+  }
+
+  switch (pjrt_buffer->element_type()) {
+    case PrimitiveType::PRED:
+    case PrimitiveType::S8:
+    case PrimitiveType::S16:
+    case PrimitiveType::S32:
+    case PrimitiveType::S64:
+    case PrimitiveType::U8:
+    case PrimitiveType::U16:
+    case PrimitiveType::U32:
+    case PrimitiveType::U64:
+    case PrimitiveType::F16:
+    case PrimitiveType::F32:
+    case PrimitiveType::F64:
+    case PrimitiveType::C64:
+    case PrimitiveType::C128:
+      break;
+
+    default:
+      throw py::attribute_error(absl::StrFormat(
+          "__cuda_array_interface__ is not supported for %s buffers.",
+          PrimitiveType_Name(pjrt_buffer->element_type())));
+  }
 
-  return IfrtHelpers::CudaArrayInterface(arr.ifrt_array(),
-                                         arr.GetStorage().dynamic_shape);
+  py::str typestr =
+      ValueOrThrow(TypeDescriptorForPrimitiveType(pjrt_buffer->element_type()));
+
+  if (!LayoutUtil::IsMonotonicWithDim0Major(pjrt_buffer->layout())) {
+    throw py::attribute_error(
+        "__cuda_array_interface__ is only supported for "
+        "buffers in row-major order.");
+  }
+
+  py::dict result;
+  const auto* dynamic_shape =
+      ValueOrThrow(IfrtHelpers::xla_dynamic_shape(ifrt_array, scratch));
+  result["shape"] = SpanToTuple(dynamic_shape->dimensions());
+  result["typestr"] = std::move(typestr);
+  std::unique_ptr<PjRtBuffer::ExternalReference> external_reference_hold =
+      ValueOrThrow(pjrt_buffer->AcquireExternalReference());
+  const void* root_ptr =
+      external_reference_hold->OpaqueDeviceMemoryDataPointer();
+  py::tuple data(2);
+  data[0] = py::int_(absl::bit_cast<std::uintptr_t>(root_ptr));
+  data[1] = py::bool_(true);  // read-only
+  result["data"] = std::move(data);
+  result["version"] = py::int_(2);
```

Comment:
In general buffers in PJRT's GPU backends are defined by CUDA events, not streams, so I think the easiest way to do this would actually be to add a stream specifically for dlpack exports and wait for the buffer definition events on that stream. The identity of that stream could then be exposed.